### PR TITLE
[WIP] Programmatically disambiguate when to step a `_LRScheduler`

### PR DESCRIPTION
**Pull request by ntenenz**
A bare-bones PoC detailing what the change proposed by #37768 would look like.

### CODE REVIEW

File: torch/optim/lr_scheduler.py [hunk 1]
```diff
@@ -234,6 +247,10 @@ def get_lr(self):
         return [base_lr * lmbda(self.last_epoch)
                 for lmbda, base_lr in zip(self.lr_lambdas, self.base_lrs)]
 
+    @property
+    def steps_on(self):
+        return StepsOnFlag.EPOCH
```
Comment (vincentqb) [inline torch/optim/lr_scheduler.py:252]: LambdaLR, and MutliplicativeLR depend on what the user wants. My impression is that we shouldn't make a choice here. The options are see here are: (1) have a way for the user to make the choice at construction, (2) let the NotImplementedError be raised, but have a way for the user to override this. Thoughts?

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -21,6 +22,13 @@
 
 SAVE_STATE_WARNING = "Please also save or load the state of the optimzer when saving or loading the scheduler."
 
+
+class StepsOnFlag(enum.Flag):
+    """A flag which indicates when a _LRScheduler should be stepped and supports standard bitwise logic."""
+    BATCH = enum.auto()  # steps at the end of a batch
+    EPOCH = enum.auto()  # steps at the end of an epoch
```
Comment (vincentqb) [inline torch/optim/lr_scheduler.py:29]: Note to self: The users would do
```python
if scheduler.steps_on() == StepsOnFlag.BATCH:
    scheduler.step()
```
```python
if scheduler.steps_on() == StepsOnFlag.EPOCH:
    scheduler.step()
```

File: torch/optim/lr_scheduler.py [hunk 3]
```diff
@@ -359,6 +380,10 @@ def _get_closed_form_lr(self):
         return [base_lr * self.gamma ** (self.last_epoch // self.step_size)
                 for base_lr in self.base_lrs]
 
+    @property
+    def steps_on(self):
+        return StepsOnFlag.EPOCH
```
Comment (vincentqb) [inline torch/optim/lr_scheduler.py:385]: Thinking out loud: Looking at `StepLR`, I could imagine that users may want to step on after a fix number of iterations, which may or may not be epochs. The property is nice in the sense that it doesn't force people to use it, yet should cover most use case.

File: torch/optim/lr_scheduler.py (hunk 2 above)
Comment (ntenenz) [reply torch/optim/lr_scheduler.py:29]: A couple clarifications:
* the method is decorated with `@property`, so no need to include the `()`
* The enum is actually a `Flag`, so using the bitwise and operator would likely be better practice

```python
for _ in range(num_epochs):
    for batch in dl:
        train_batch(batch)
        if scheduler.steps_on & StepsOnFlag.BATCH:
            scheduler.step()

    validate()
    if scheduler.steps_on & StepsOnFlag.EPOCH:
        scheduler.step()
```

File: torch/optim/lr_scheduler.py (hunk 3 above)
Comment (ntenenz) [reply torch/optim/lr_scheduler.py:385]: As discussed in the issue, a few of the schedulers could be leveraged by users in less traditional manners (e.g., every B batches, every E epochs, or based on some conditional logic). As is, I've currently set them to signal the stepping behavior as described in the documentation (at least as of PyTorch 1.4).

For the default schedulers, there's not much we can do other than let the user maintain the non-traditional stepping logic in their codebase and let them ignore the flag. However, if a user is to create their own scheduler, either by inheriting from an existing one or creating their own, they can dynamically set the `steps_on` property using arbitrary internal logic to ensure that code which follows the generic pattern listed below would still work. As an aside, if we wish to enable users to more easily support this use case, we may wish to include a `StepsOnFlag.NONE` to indicate that it should neither step on a batch nor epoch for the time being.

```python
for _ in range(num_epochs):
    for batch in dl:
        train_batch(batch)
        if scheduler.steps_on & StepsOnFlag.BATCH:
            scheduler.step()

    validate()
    if scheduler.steps_on & StepsOnFlag.EPOCH:
        scheduler.step()
```

File: torch/optim/lr_scheduler.py (hunk 1 above)
Comment (vincentqb) [reply torch/optim/lr_scheduler.py:253]: see also [comment](https://github.com/pytorch/pytorch/issues/37768#issuecomment-659536888)

Comment (ntenenz) [reply torch/optim/lr_scheduler.py:253]: The motivation for this is described [here](https://github.com/pytorch/pytorch/issues/37768#issuecomment-659365012) and [here](https://github.com/pytorch/pytorch/issues/37768#issuecomment-659564127). Namely, the decision seems to come down to whether standard/default behavior should be provided. To ensure backwards compatibility (given this likely will not land in a major release), I'd argue it should and that it should follow existing documentation, even if the described behavior is not enforced.

It certainly seems reasonable to allow the user to override this flag (upon construction, via inheritance, etc.), however, I wanted to minimize the scope of the functionality added in this PR. Given that both an extra `__init__` param or redefined behavior via inheritance would both be backwards compatible with this change (and the current API), I didn't feel particularly compelled to include it here. Happy to fold it in though if you'd like.
